include/mlir/Dialect/SPIRV/SPIRVOps.td

```diff
@@ -172,6 +172,78 @@ def SPV_CompositeExtractOp : SPV_Op<"CompositeExtract", [NoSideEffect]> {
 
 // -----
 
+def SPV_ControlBarrierOp : SPV_Op<"ControlBarrier", []> {
+  let summary = [{
+    Wait for other invocations of this module to reach the current point of
+    execution.
+  }];
+
+  let description = [{
+    All invocations of this module within Execution scope must reach this
+    point of execution before any invocation will proceed beyond it.
+
+    When Execution is Workgroup or larger, behavior is undefined if this
+    instruction is used in control flow that is non-uniform within
+    Execution. When Execution is Subgroup or Invocation, the behavior of
+    this instruction in non-uniform control flow is defined by the client
+    API.
+
+    If Semantics is not None, this instruction also serves as an
+    OpMemoryBarrier instruction, and must also perform and adhere to the
+    description and semantics of an OpMemoryBarrier instruction with the
+    same Memory and Semantics operands.  This allows atomically specifying
+    both a control barrier and a memory barrier (that is, without needing
+    two instructions). If Semantics is None, Memory is ignored.
+
+    Before version 1.3, it is only valid to use this instruction with
+    TessellationControl, GLCompute, or Kernel execution models. There is no
+    such restriction starting with version 1.3.
+
+    When used with the TessellationControl execution model, it also
+    implicitly synchronizes the Output Storage Class:  Writes to Output
+    variables performed by any invocation executed prior to a
+    OpControlBarrier will be visible to any other invocation after return
+    from that OpControlBarrier.
+
+    ### Custom assembly form
+
+    ``` {.ebnf}
+    scope ::= `"CrossDevice"` | `"Device"` | `"Workgroup"` | `"Subgroup"`
+              | `"Invocation"`
+
+    memory-semantics ::= `"None"` | `"Acquire"` | `"Release"`
+                         | `"AcquireReleae"` | `"SequentiallyConsistent"`
+                         | `"UniformMemory"`  | `"SubgroupMemory"`
+                         | `"WorkgroupMemory"`  | `"CrossWorkgroupMemory"`
+                         | `"AtomicCounterMemory"` | `"ImageMemory"`
+
+    control-barrier-op ::= `spv.ControlBarrier` scope, scope, memory-semantics
+    ```
+
+    For example:
+
+    ```
+    spv.ControlBarrier "Workgroup", "Device", "Acquire|UniformMemory"
+
+    ```
+  }];
+
+  let arguments = (ins
+    SPV_ExecutionScopeAttr:$execution_scope,
+    SPV_MemoryScopeAttr:$memory_scope,
+    SPV_MemorySemanticsAttr:$memory_semantics
+  );
+
+  let results = (outs);
+
+  let verifier = [{ return verifyMemorySemantics(*this); }];
```

Comment:
I think this is just the default and can be removed?

Reply:
@antiagainst can you please comment, do I get you right? I have to make `verifyMemorySemantics` code to be autogenerated?

Reply:
Nvm, this is fine.  Previously I meant that if we delete this line here, because of the SPV_Op definition, we will need to supply a `verify()` method for `spv::ControlBarrierOp` anyway. But that means `verify()` needs to  wrap `verifyMemorySemantics()` here.